### ISSUE TITLE
examples/char_prediction cost is nan

### ISSUE THREAD

**Issue by byzhang** [JonathanRaiman/Dali#6]
when run 
```
examples/character_prediction
```
When running a while, the softmax_cross_entropy_rowwise() returns nan.

I added some log, and found after https://github.com/JonathanRaiman/Dali/blob/master/dali/tensor/op/cost.cu.cpp#L363, the out.w(0) is 0, so log(out) become a Nan. 

@JonathanRaiman , did you see the issue before? I'm using all default flag values.

**Comment by JonathanRaiman** [comment]
Did not run into this problem before; I believe this may happen on GPU more often than on CPU due to numerical imprecisions. You can change `softmax_cross_entropy_rowwise(logprobs, labels)` to `cross_entropy(softmax(logprobs) + eps, labels)` with eps = 1e-6.

**Comment by byzhang** [comment]
How about this change on https://github.com/JonathanRaiman/Dali/blob/master/dali/tensor/op/cost.cu.cpp#L229
from
```
MAT(out) = (R)-1.0 * F<op::log<R>>(MAT(out).wrapper());
```
to
```
MAT(out) = (R)-1.0 * F<op::log<R>>(MAT(MatOps<R>::max(out, std::numeric_limits<R>::min())).wrapper());
```
So it won't affect GRAD(out), and still compute softmax and cross_entropy in one layer.
But I'm not sure whether any other side-effects of this change.
